mito-ai/src/Extensions/InlineCompleter/provider.ts

```diff
@@ -0,0 +1,403 @@
+        return {
+          items: []
+        };
+      }
+      const mime = request.mimeType ?? IEditorMimeTypeService.defaultMimeType;
+      const language = this._languageRegistry.findByMIME(mime);
+      if (!language) {
+        console.warn(
+          `Could not recognize language for ${mime} - cannot complete`
+        );
+        return { items: [] };
+      }
+
+      let cellId = undefined;
+      let path = context.session?.path;
+      if (context.widget instanceof NotebookPanel) {
+        const activeCell = context.widget.content.activeCell;
+        if (activeCell) {
+          cellId = activeCell.model.id;
+        }
+      }
+      if (!path && context.widget instanceof DocumentWidget) {
+        path = context.widget.context.path;
+      }
+      const messageId = ++this._counter;
+
+      const stream = true;
+
+      const prefix = this._getPrefix(request);
+      const result = await this._client.sendMessage({
+        path: context.session?.path,
+        mime,
+        prefix,
+        suffix: this._getSuffix(request),
+        language: this._resolveLanguage(language),
+        message_id: messageId.toString(),
+        stream,
+        cell_id: cellId
+      });
+
+      this._currentPrefix = prefix;
+      for (let index = prefix.length - 1; index >= 0; index--) {
+        if (prefix[index] === '\n') {
+          this._currentPrefix = prefix.slice(index + 1);
+          break;
+        }
+      }
+      if (stream && result.list.items[0]?.token) {
+        this._currentToken = result.list.items[0].token;
+        this._currentStream = new Stream<
+          MitoAIInlineCompleter,
+          InlineCompletionStreamChunk
+        >(this);
+      }
+
+      const error = result.error;
+      if (error) {
+        this._notifyCompletionFailure(error);
+        throw new Error(
+          `Inline completion failed: ${error.type}\n${error.traceback}`
+        );
+      }
+
+      return result.list;
+    } finally {
+      this._completionLock.resolve();
+    }
+  }
+
+  /**
+   * Whether the completer is enabled or not.
+   */
+  isEnabled(): boolean {
+    return this._settings.enabled;
+  }
+
+  /**
+   * Stream a reply for completion identified by given `token`.
+   */
+  async *stream(token: string): AsyncGenerator<{
+    response: IInlineCompletionItem;
+  }> {
+    if (!this.isEnabled()) {
+      throw new Error('Mito AI completion is disabled.');
+    }
+    // Wait for the acknowledge request to be fulfilled before starting the stream
+    await this._completionLock.promise;
+
+    if (this._currentToken !== token) {
+      // New completion may be triggered before the code have started streaming
+      // the previous one. So not raising an error here.
+      console.debug(`No stream found for token '${token}'.`);
+      return;
+    }
+
+    for await (const chunk of this._currentStream!) {
+      if (this._currentToken !== token) {
+        break;
+      }
```

Comment:
Actually it may happen. If JupyterLab has started consuming that stream and a new request comes in, the `_currentToken` will change and we stop the stream has useless has the previously completion should be discarded.